File: csharp/resistor-color-trio/ResistorColorTrio.cs
using System;
using System.Collections.Generic;

public static class ResistorColorTrio
{
    private const int KILO = 1000;
    private static readonly IReadOnlyDictionary<string, int> s_resistorColors;
    static ResistorColorTrio() => s_resistorColors = new Dictionary<string, int>()
    {
        ["black"] = 0,
        ["brown"] = 1,
        ["red"] = 2,
        ["orange"] = 3,
        ["yellow"] = 4,
        ["green"] = 5,
        ["blue"] = 6,
        ["violet"] = 7,
        ["grey"] = 8,
        ["white"] = 9,
    };

    public static string Label(string[] colors)
    {
        if (colors == null || colors.Length != 3)
            throw new ArgumentException($"{nameof(colors)} does not have 3 resistanceValues.");
            
        var resistanceValue = CalculateResistance(ResistanceValues(colors));
        (string prefix, resistanceValue) = resistanceValue > KILO 
            ? ("kilo", resistanceValue / KILO) 
            : (string.Empty, resistanceValue);

        return $"{resistanceValue} {prefix}ohms";
    }

    private static int[] ResistanceValues(string[] colors)
    {
        int[] resistanceValues = new int[colors.Length];
        for (int i = 0; i < colors.Length; i++)
            if (!s_resistorColors.TryGetValue(colors[i], out resistanceValues[i]))
                throw new ArgumentException($"'{colors[i]}' is not a resistor color.");

        return resistanceValues;
    }

    private static int CalculateResistance(int[] resistanceValues) 
        => (resistanceValues[0] * 10 + resistanceValues[1]) * (int)Math.Pow(10, resistanceValues[2]);
}

File: go/resistor-color-trio/resistor_color_trio.go
// Package resistorcolortrio calculates and formats a label for a resistor
// of three color bands.
package resistorcolortrio

import (
	"fmt"
	"math"
)

var prefixes = [...]string{"", "kilo", "mega", "giga"}

// Label describes the resistance value given the colors of a resistor.
// The label is a string with a resistance value with an unit appended
// (e.g. "33 ohms", "470 kiloohms").
func Label(colors []string) string {
	if len(colors) < 3 {
		return ""
	}

	value := resistorColors[colors[0]]*10 + resistorColors[colors[1]]
	value *= int(math.Pow10(resistorColors[colors[2]]))
	j := 0
	for ; j < len(prefixes) && value > 1_000; value /= 1_000 {
		j++
	}

	return fmt.Sprintf("%d %sohms", value, prefixes[j])
}

var resistorColors = map[string]int{
	"black":  0,
	"brown":  1,
	"red":    2,
	"orange": 3,
	"yellow": 4,
	"green":  5,
	"blue":   6,
	"violet": 7,
	"grey":   8,
	"white":  9,
}

// BenchmarkLabel-4          535584              2082 ns/op             248 B/op         19 allocs/op
